iterators.rb

```diff
@@ -0,0 +1,132 @@
+# rubocop:disable  Metrics/ModuleLength, Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+module Enumerable
+  def my_each
+    index = 0
+    while index < size
+
+      case self.class.name
+      when 'Hash' then yield(keys[index], self[keys[index]])
+      when 'Array' then yield(self[index])
+      when 'Range' then yield(to_a[index])
+      end
+      index += 1
+    end
+  end
+
+  def my_each_with_index
+    index = 0
+    while index < size
+      yield(self[index], index)
+      index += 1
+    end
+  end
+
+  def my_select
+    return enum_for unless block_given?
+
+    array = []
+    my_each do |element|
+      array.push(element) if yield(element) == true
+    end
+    array
+  end
+  alias my_find_all my_select
+  alias my_filter my_select
+
+  def my_all?(pattern = nil)
+    is_equal = true
+    if pattern.nil?
+      if block_given?
+        my_each { |val| is_equal = false unless yield(val) or size.zero? }
+      else my_each { |val| is_equal = false unless val == true }
+      end
+    elsif pattern.is_a?(Module)
+      if pattern.is_a?(Regexp)
+        my_each { |val| is_equal = false unless pattern.match(val) }
+      else
+        my_each { |val| is_equal = false unless val.is_a?(pattern) }
+      end
+    end
+    is_equal
+  end
+
+  def my_any?
+    my_each do |val|
+      return true if yield(val)
+    end
+    false
+  end
+
+  def my_none?(pattern = nil)
+    is_true = true
+    if pattern.nil?
+      if block_given?
+        my_each { |val| is_true = false if yield(val) or size.zero? }
+      else my_each { |val| is_true = false if val == true }
+      end
+    elsif pattern.is_a?(Module)
+      if pattern.is_a?(Regexp)
+        my_each { |val| is_true = false if pattern.match(val) }
+      else
+        my_each { |val| is_true = false if val.is_a?(pattern) }
+      end
+    end
+    is_true
+  end
```

Comment:
- [ ] should have working `my_none?` identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

#### my_none?

- [ ] when a Regex is passed as an argument returns true only if none of the collection matches the Regex
```ruby
%w[dog door rod blade].my_none?(/o/) == %w[dog door rod blade].none?(/o/) => false
```
- Yours is retrieving the wrong answer
![image](https://user-images.githubusercontent.com/52209067/80292854-213f1000-8717-11ea-980e-8240aab70326.png)

- Real none?
![image](https://user-images.githubusercontent.com/52209067/80292860-3156ef80-8717-11ea-9805-0cf7c33b1876.png)

- [ ] when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
```ruby
[3,4,7,11].my_none?(3) == [3,4,7,11].none?(3) => false
```
- Yours is retrieving the wrong answer
![image](https://user-images.githubusercontent.com/52209067/80292878-53e90880-8717-11ea-8222-11ab6c341652.png)

- Real none?
![image](https://user-images.githubusercontent.com/52209067/77690915-a9d85e00-6f69-11ea-81b3-7adb07b281ca.png)

[OPTIONAL]

I think we could get the block of my_none? from my_any? we should only do something like !my_any? for my_none? to work. For example:

  def my_none?(arg = nil, &block)
    !my_any?(arg, &block)
  end

To make this work we have to use block.call(element) instead of yield(element) in my_any?

Reply:
**Fixed**
I have considered your optional suggestion  :thinking:  to reuse the code already written inside my_any? since they are opposite of each other.
However, the problem I have with that approach is, first, the original one doesn't take a block as a second argument. But that doesn't mean we can not improve the original one :wink: which brings me to the second and main problem, if we make both methods depend on each other the flexibility and cohesion will be lost.